addons/sale_project/models/sale_order_line.py

```diff
@@ -58,16 +58,43 @@ def default_get(self, fields):
                     sale_order = self.env['sale.order'].create(so_create_values)
                     sale_order.action_confirm()
                 default_values['order_id'] = sale_order.id
-            if (name := self.env.context.get('default_name')):
+            if product_name := self.env.context.get('sol_product_name') or self.env.context.get('default_name'):
                 product = self.env['product.product'].search([
-                    ('name', 'ilike', name),
+                    ('name', 'ilike', product_name),
                     ('company_id', 'in', [False, self.env.company.id]),
                 ], limit=1)
                 if product:
                     default_values['product_id'] = product.id
+                    # Avoid overwriting the computed name if we find a product.
+                    del default_values['name']
+                    # We also need to remove the name from the defaults
+                    # so that the name of the SOL is based on the full name of
+                    # the product and not just what was typed in the field.
+                    if "name" in res:
+                        del res["name"]
             return {**res, **default_values}
         return res
 
+    @api.model
+    def name_create(self, name):
+        # To get the right product when creating a SOL on the fly, we need to get
+        # the name that was entered in the field from the `default_get` method.
+        # The easiest way of doing that is to store it in the context.
+        if self.env.context.get('form_view_ref') == 'sale_project.sale_order_line_view_form_editable':
+            self = self.with_context(sol_product_name=name)
+        return super().name_create(name)
+    
+    @api.model
+    def _add_missing_default_values(self, values):
+        # When creating a SOL through the quick create, the name_create will be
+        # called with whatever was typed in the field. However, we don't want
+        # that value to overwrite the computed SOL name if we find a product.
+        defaults = super()._add_missing_default_values(values)
+        if self.env.context.get('form_view_ref') == 'sale_project.sale_order_line_view_form_editable':
+            if "name" in defaults and "product_id" in defaults:
+                del defaults["name"]
+        return defaults
```

Comment:
I shouldn't be overriding this, but I don't know how else I could avoid the name typed in the field from being used in the `create`.